src/daft-logical-plan/src/stats.rs

```diff
@@ -0,0 +1,121 @@
+use std::fmt::Display;
+
+use serde::{Deserialize, Serialize};
+
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq, Hash)]
+pub enum StatsState {
+    Materialized(PlanStats),
+    NotMaterialized,
+}
+
+impl StatsState {
+    pub fn unwrap_or_default(&self) -> PlanStats {
+        match self {
+            Self::Materialized(stats) => stats.clone(),
+            Self::NotMaterialized => PlanStats::default(),
+        }
+    }
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct PlanStats {
+    // Currently we're only putting cardinality stats in the plan stats.
+    // In the future we want to start including column stats, including min, max, NDVs, etc.
+    pub approx_stats: ApproxStats,
+}
+
+impl PlanStats {
+    pub fn new(approx_stats: ApproxStats) -> Self {
+        Self { approx_stats }
+    }
+
+    pub fn empty() -> Self {
+        Self {
+            approx_stats: ApproxStats::empty(),
+        }
+    }
+}
+
+impl Default for PlanStats {
+    fn default() -> Self {
+        Self::empty()
+    }
+}
+
+impl Display for PlanStats {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "{{ Lower bound rows = {}, Upper bound rows = {}, Lower bound bytes = {}, Upper bound bytes = {} }}",
+            self.approx_stats.lower_bound_rows,
+            self.approx_stats.upper_bound_rows.map_or("None".to_string(), |v| v.to_string()),
+            self.approx_stats.lower_bound_bytes,
+            self.approx_stats.upper_bound_bytes.map_or("None".to_string(), |v| v.to_string()),
+        )
+    }
+}
+
+// We implement PartialEq, Eq, and Hash so that all PlanStats are considered equal. This allows
+// logical/physical plans that are enriched with stats to easily implement PartialEq, Eq, and Hash
+// in a way that ignores PlanStats when considering equality.
```

Comment:
this looks pretty sketch. We should hash, eq, partial_eq it properly here. and handle this at the plan.

An idea of how to do that is to use the new type pattern.

```
struct AlwaySame<T>(T);

impl<T> Deref for AlwaySame<T> ..

impl<T> Hash for AlwaySame<T> {
    #[inline]
    fn hash<H: std::hash::Hasher>(&self, _state: &mut H) {
        // Add nothing to hash state since all PlanStats should hash the same.
    }
}

impl<T> PartialEq for AlwaySame<T> ...
```

and then in the logical plan hold onto `AlwaySame<PlanStats>`

Reply:
Sounds good, done